docs/adr/0010-mise-en-place-de-bounded-contexts.md

```diff
@@ -0,0 +1,176 @@
+
+Date : 2020-04-02
+
+## État
+
+Draft
+
+## Contexte
+
+> les organisations qui conçoivent des systèmes [...] sont contraintes de produire des designs qui sont des copies de la structure de communication de leur organisation.
+> — M. Conway
+
+L'architecture logicielle de Pix API a été pensée et initiée il y a 3 ans. 
+
+Depuis, beaucoup de choses ont changé :
+
+- **métier** : de nouvelles activités stratégiques sont apparues (évaluation, gestion des comptes & accès, prescription, certification, administration)
+- **organisation** : nous sommes passés d'1 équipe de 6 personnes à 4 équipes produit de 6 personnes
+- **technique** : les technologies et leurs fonctionnalités ont évoluées de même que notre compréhension et notre maîtrise de celles-ci ainsi que de certains concepts (clean architecture, architecture hexagonale, etc.)
+- **culturelle** : les développeurs de cette époque ont quitté le projet
+
+En regard de tous ces changements, nos patterns, nos pratiques et nos standards ont finalement peu évolués. Ce qui nous pose de plus en plus de problèmes :
+
+- du fait du départ de leur porteur principal, certaines initiatives n'ont pas été jusqu'au bout (ex : passage des Services aux Use cases, mise en ouvre du pattern RORO, etc.)
+- les objets du Domaine comme les objets de Response sont désormais des objets obèses, fourre-tout, potentiellement incomplets ou inintelligents et le plus souvent sous-utilisés car multi-usages
+- les standards et règles d'implémentation, qui s'accomodaient bien d'un fonctionnement en comité restreint, s'avèrent gênantes ou contre-productives, quand elles ne sont pas tout simplement oubliées
+- il en va de même pour le design de l'API, qui cumule les conséquences néfastes des 2 points précédents
+
+Enfin, le succès de Pix et le développement de ses partenariats font émerger de nouveaux besoins que l'API, sous sa forme actuelle, risque d'empêcher ou limiter : 
+
+- interconnexion multiple et diverse avec des SI partenaires / clients / fournisseurs
+- caractère open-contribution de la plateforme afin de favoriser sont adoption par le plus grand nombre (dont l'international), le recrutement de collaborateurs ou l'émergence d'innovation
+
+Tous ces freins, douleurs, risques ou opportunités nous poussent à croire qu'il est aujourd'hui temps pour Pix de dépoussiérer son API et de revoir son architecture.
+
+## Décision
+
+Nous prenons la décision de réorganiser le code source de l'application Pix API en contextes métier (_Bounded Contexts_, a.k.a. BC).
+
+La mise en œuvre des BC suit les principes directeurs suivants :
+1. **Architecture hexagonale** à 3 couches plutôt que clean architecture à 4 couches
+1. Nommage des contextes par **finalité métier** plutôt que par application cible
+1. Matérialisation de ces contextextes par **modules de code**
+1. Utilisation de **plugins Hapi.js** pour matérialiser ces modules de code
+1. **Duplication de circonstance** pour les données utiles à plusieurs contextes
```

Comment:
Je me demande si les lignes proposées délimitent bien des _bounded contexts_ (j'ai un doute au passage sur la traduction "contexte métier" qui me semble perdre de l'information). Cet article par exemple : https://medium.com/naxex/bounded-contexts-are-not-microservices-ead44b8d6e35 propose une formulation (indépendamment de l'aspect "microservices") que je trouve intéressante :
> A Bounded Context defines the boundaries of the biggest services possible: services that won’t have any conflicting models inside of them.

De ce point de vue, j'ai l'impression qu'il y a peu de conflits dans le langage utilisé à l'intérieur de Pix. Pour prendre directement le cas le plus épineux : un "utilisateur de Pix" désigne il me semble le même concept du point de vue évaluation, prescription ou certification. En ce qui concerne le langage du métier, en écoutant les gens des différents pôles échanger, je n'ai pas l'impression qu'il y ait de conflits entre leurs utilisations des mêmes termes — comme il peut y en avoir par exemple entre la notion de "produit" pour quelqu'un qui gère un catalogue, qui est différente du "produit" pour la personne qui gère le stock et encore pour celle qui développe le "produit" en question.

En revanche, on a constaté que dans le code, on ressent le besoin d'avoir des vues différentes des mêmes entités. Mais ces besoins différents ne s'alignent même pas sur les finalités métier : pour rester sur le cas de "l'utilisateur", dans le contexte métier "prescription" on ne veut pas la même vue pour alimenter une liste de 3000 prescrits que pour afficher les détails d'un prescrit — et pourtant c'est la même finalité métier.

En résumé, je ne suis pas (encore ?) convaincu que les _bounded contexts_ soient une ligne de découpage pertinente ici. Par ailleurs, j'ai l'impression qu'un découpage en _bounded contexts_ serait contradictoire avec le fait de continuer à partager un stockage unique (table `users` par exemple).